### PR TITLE
Changed map to filter map so that Phragmen ignores empty voters

### LINKED ISSUE #7172: Phragmen should ignore "empty" voters
We build the voters as such: 

https://github.com/paritytech/substrate/blob/0c8329c71c241cce09feb0372ee5d5d4a161d11a/primitives/npos-elections/src/lib.rs#L385-L405

If by the end of the `map()`, `edges.len()` is empty, then this voter should not be even included. Essentially, we have to change the `map()` to `filter_map`. 

Ideally there should be a test added for this, else we need to make sure the change is reflected in #6685 as well. 

This (as far as I know) does not cause any issues. It is just extra iterations here and there that will make the execution a bit slower.

### PR DESCRIPTION
**Pull request by ksr30**
Fixed #7172

### CODE REVIEW

File: primitives/npos-elections/src/lib.rs [hunk 1]
```diff
@@ -629,32 +629,37 @@ pub(crate) fn setup_inputs<AccountId: IdentifierT>(
...
-					candidate.approval_stake.saturating_add(voter_stake.into());
-				edges.push(
-					Edge {
-						who: v.clone(),
-						candidate: Rc::clone(&candidates[*idx]),
-						..Default::default()
-					}
-				);
-			} // else {} would be wrong votes. We don't really care about it.
+		if edges.is_empty() {
```
Comment (kianenigma) [inline primitives/npos-elections/src/lib.rs:634]: This is always going to be empty! you just build a new `edges` and you immediately assert that it is empty. 

You can't do this like such. You need to go over all `voters`, some might be duplicate, some might also be invalid  
(`// else {} would be wrong votes. We don't really care about it.`). Once all is done, before returning `Voter { .. } ` you check if it has any edges or not. 

Also, a test would be good.

**Review by kianenigma — CHANGES_REQUESTED** [review]
Has some flaws.

File: primitives/npos-elections/src/lib.rs (hunk 1 above)
Comment (ksr30) [reply primitives/npos-elections/src/lib.rs:634]: Okay ... got it

**Comment by bkchr** [comment]
@kianenigma so what do we want to do?

**Comment by kianenigma** [comment]
Waiting for @ksr30 to address the comment. Can take over it later myself as well.

**Comment by ksr30** [comment]
Hey @kianenigma , sorry for late reply ... I will work on this tomorrow

**Comment by ksr30** [comment]
Hey @kianenigma , I pushed required changes ... please have a look

**Comment by ksr30** [comment]
Hey @kianenigma , is everything fine with PR. I see few checks are failing ... but maybe its because of nightly version. Otherwise all tests are passing on my system.

**Comment by ksr30** [comment]
Hey @kianenigma ,sorry to disturb you again ... is everything good with PR or you want me to change anything?

**Comment by bkchr** [comment]
@ksr30 as a start, please merge master.

**Comment by ksr30** [comment]
Hey @bkchr .. sorry i didnt get

**Comment by bkchr** [comment]
Merge master into your branch.

File: primitives/npos-elections/src/tests.rs [hunk 2]
```diff
@@ -72,6 +72,50 @@ fn float_phragmen_poc_works() {
 	);
 }
 
+#[test]
+fn phragmen_core_setup_inputs() {
```
Comment (kianenigma) [inline primitives/npos-elections/src/tests.rs:76]: This test case is not testing the change of this PR. The logic of this PR is to see what happens when there is a voter with no edges. You are not checking this here.

**Review by kianenigma — CHANGES_REQUESTED** [review]
Please write a test case that actually tests the logic change of the PR.

File: primitives/npos-elections/src/tests.rs (hunk 2 above)
Comment (ksr30) [reply primitives/npos-elections/src/tests.rs:76]: (10, 10, vec![1, 2]) 
@kianenigma you mean check with blank vector. something like this (10,10, vec![]). Right?

**Comment by ksr30** [comment]
@kianenigma  @bkchr sorry for delay ... please have a look

File: primitives/npos-elections/src/tests.rs [hunk 3]
```diff
@@ -72,6 +72,46 @@ fn float_phragmen_poc_works() {
 	);
 }
 
+#[test]
+fn phragmen_core_test_without_edges() {
+	let candidates = vec![1, 2, 3];
+	let voters = vec![
+		(10, 10, vec![]),
```
Comment (thiolliere) [inline primitives/npos-elections/src/tests.rs:79]: we could also test for non candidate vote here
```suggestion
		(10, 10, vec![4, 5, 6]),
```

File: primitives/npos-elections/src/lib.rs [hunk 4]
```diff
@@ -629,7 +629,7 @@ pub(crate) fn setup_inputs<AccountId: IdentifierT>(
 		})
 		.collect::<Vec<CandidatePtr<AccountId>>>();
 
-	let voters = initial_voters.into_iter().map(|(who, voter_stake, votes)| {
+	let voters = initial_voters.into_iter().filter_map(|(who, voter_stake, votes)| {
```
Comment (thiolliere) [inline primitives/npos-elections/src/lib.rs:632]: doc of function could be changed to include this change.

File: primitives/npos-elections/src/lib.rs [hunk 5]
```diff
@@ -650,12 +650,18 @@ pub(crate) fn setup_inputs<AccountId: IdentifierT>(
...
 		}
-		Voter {
-			who,
-			edges: edges,
-			budget: voter_stake.into(),
-			load: Rational128::zero(),
+		if edges.is_empty() {
+			None
+		}
+		else {
```
Comment (thiolliere) [inline primitives/npos-elections/src/lib.rs:656]: I think this is the standard format in substrate
```suggestion
		} else {
```

**Review by thiolliere — APPROVED** [review]
as far as I see, removing empty voters doesn't harm phragmms nor seq_phragmen

**Comment by ksr30** [comment]
Thanks @thiolliere

**Comment by ksr30** [comment]
@kianenigma @bkchr if PR is okay ... then can we merge it now :sweat_smile:

**Review by kianenigma — APPROVED** [review]

**Comment by thiolliere** [comment]
@ksr30 can you merge master again ? so that CI is happy

**Comment by kianenigma** [comment]
@ksr30 ping

**Comment by ksr30** [comment]
> @ksr30 ping

Sorry @kianenigma  missed the notification ... I will merge master by tomorrow

**Comment by ksr30** [comment]
@kianenigma @thiolliere ... I think we can merge it now

**Comment by thiolliere** [comment]
bot merge

**Pull request merged**